### PR TITLE
fix FakeTensor creation on noncontiguous subclasses

### PR DESCRIPTION
**Pull request by bdhirsh**
Fixes https://github.com/pytorch/pytorch/issues/125287

Fixes https://github.com/pytorch/pytorch/issues/124090, context on the issue

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #125288
* #124400
* __->__ #124399
* #124398

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @fegin @XilunWu @wanchaol @fduwjj @wz337 @tianyu-l @wconstab @yf225 @chauhang @d4l3k @ezyang @msaroufim @anijain2305 @voznesenskym @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng

### CODE REVIEW

File: torch/autograd/graph.py [hunk 1]
```diff
@@ -203,6 +204,21 @@ def increment_version(tensor):
...
+    This effectively forbids the tensor from having a gradient computed during backward.
+    If the tensor was originally an autograd leaf (tensor.is_leaf == False),
+    setting the tensor's grad_fn to an error node will flip tensor.is_leaf to True.
+
+    This is a convenient API used in torch.compile internals, when we need to
+    take a graph input (potentially a non-leaf), and generate a fresh FakeTensor
+    for it: we need the FakeTensor to have accurate is_leaf information,
+    even though we don't actually plan to run autograd through the graph input.
+    """
+    torch._C._forbid_in_autograd(tensor)
```
Comment (bdhirsh) [inline torch/autograd/graph.py:219]: @albanD does this look ok to you as public API + the docs I wrote? Let me know if you think I should make it clearer  that it's a sharp-edge API

File: torch/csrc/autograd/variable.cpp [hunk 2]
```diff
@@ -363,6 +364,16 @@ void bump_version(const Variable& self) {
   self.unsafeGetTensorImpl()->bump_version();
 }
 
+void forbid_in_autograd(const Variable& self) {
+  TORCH_CHECK(
```
Comment (albanD) [inline torch/csrc/autograd/variable.cpp:368]: You need a bit more error checking here to ensure that:
- This is a leaf (no existing grad_fn)
- This is not already part of a graph (no grad_accumulator)

File: torch/csrc/autograd/variable.cpp [hunk 3]
```diff
@@ -363,6 +364,16 @@ void bump_version(const Variable& self) {
...
+      self.defined(), "cannot call forbid_in_autograd() on undefined tensor");
+  auto new_grad_fn = std::shared_ptr<torch::autograd::Error>(
+      new torch::autograd::Error(
+          "Cannot backprop through Error node, file a bug in PyTorch"),
```
Comment (albanD) [inline torch/csrc/autograd/variable.cpp:372]: Cannot backprop through a Tensor that was marked as forbidden in backward.
Or something similar

File: torch/autograd/graph.py [hunk 4]
```diff
@@ -203,6 +204,21 @@ def increment_version(tensor):
     torch._C._increment_version(tensor)
 
 
+def forbid_in_autograd(tensor):
```
Comment (albanD) [inline torch/autograd/graph.py:207]: forbid_in_backward() ?
It's not forbidden in forward mode AD or from interracting in a non-differentiable way with autograd.

File: torch/autograd/graph.py [hunk 5]
```diff
@@ -203,6 +204,21 @@ def increment_version(tensor):
...
+    """Replaces the current tensor's grad_fn with an Error node.
```
Comment (albanD) [inline torch/autograd/graph.py:208]: While I like the factuality and it makes it clear to me what this does, I think we need a bit more sugar coating for end users. See below.

Also maybe have a small note that this is an advanced API that we don't expect most users to use and we expect that detach() and no_grad() should be used by most users to locally disable autograd as discussed in https://pytorch.org/docs/stable/notes/autograd.html#locally-disabling-gradient-computation

File: torch/autograd/graph.py [hunk 6]
```diff
@@ -203,6 +204,21 @@ def increment_version(tensor):
...
+
+    This effectively forbids the tensor from having a gradient computed during backward.
+    If the tensor was originally an autograd leaf (tensor.is_leaf == False),
+    setting the tensor's grad_fn to an error node will flip tensor.is_leaf to True.
```
Comment (albanD) [inline torch/autograd/graph.py:212]: You inverted the True/False here

File: torch/autograd/graph.py [hunk 7]
```diff
@@ -203,6 +204,21 @@ def increment_version(tensor):
     torch._C._increment_version(tensor)
 
 
+def forbid_in_autograd(tensor):
+    """Replaces the current tensor's grad_fn with an Error node.
+
+    This effectively forbids the tensor from having a gradient computed during backward.
```
Comment (albanD) [inline torch/autograd/graph.py:210]: This sounds like the right intro for just above

File: torch/autograd/graph.py [hunk 8]
```diff
@@ -203,6 +204,21 @@ def increment_version(tensor):
...
+    If the tensor was originally an autograd leaf (tensor.is_leaf == False),
+    setting the tensor's grad_fn to an error node will flip tensor.is_leaf to True.
+
+    This is a convenient API used in torch.compile internals, when we need to
```
Comment (albanD) [inline torch/autograd/graph.py:214]: .. note:: One example where this API is used is ...

**Review by albanD — COMMENTED** [review]
That sounds fair as a public API for me. I'll let @soulitzer give his opinion in case he prefers to keep it private.
Either way, it needs testing in test_autograd!

**Comment by soulitzer** [comment]
Public API sounds okay, but preferably with a clear story around when we'd use this over existing APIs that do similar "forbid_in_autograd" things like `.detach()` or setting `.requires_grad=False`. (We also have the private torch._C._functions.DelayedError, which is an out-of-place version of this.) If it is difficult to have such examples, I don't mind keeping it private either, but no strong preference.

**Comment by bdhirsh** [comment]
I updated the PR with @soulitzer's idea to not add a new API: it seems like `torch._C._functions.DelayedError` should do the job.

Adding that ErrorNode also caused some test failures, which made me realize that `autograd.backward` is broken in dynamo: https://github.com/pytorch/pytorch/issues/125287. I just fixed it directly in this PR.

**Review by soulitzer — APPROVED** [review]

**Comment by bdhirsh** [comment]
Hmm... the errors I'm getting are because `y = torch._C._functions.DelayedError(1)(x)` don't "propagate" the fact that x is a fake tensor (it returns a plain tensor)
